### ISSUE TITLE
Image Compression Takes Three Passes

### ISSUE THREAD

**Issue by mjkeating** [apache/incubator-pagespeed-mod#1171]
We are  using Pagespeed as a forward proxy on Apache 2.4.  

Anyway, unless already cached, all images consistently take three GET attempts, through our Pagespeed enabled proxy, before coming back compressed.  I tried increasing the ModPagespeedRewriteDeadlinePerFlushMs setting (even setting it to infinite), to no avail.

Is there anything I can do to get back compressed images on the first try?

Many thanks.

**Comment by jmarantz** [comment]
If you are using OptimizeForBandwidth (or similar options) then that's
(currently) the way it works:
1.  One pass for collecting the image bytes into our cache
2.  One pass for pulling it from the cache and initiating a background
   image rewrite & storing the optimized version in the cache
   Currently there's a bug (limitation) that the deadline you configured does
   not apply on in-place rewrites.  I have actually implemented a fix but have
   not had enough time to land it.  But that would just get you to two passes
   even if it worked.

What _does_ work today is if your images were stored on the local disk at
the server and loaded via LoadFromFile.  However that is probably not what
you want as a forward proxy.

However if you are letting us change img URLs in HTML files (CoreFilters
etc) then it should work in two passes, or even on the first pass if your
deadline is long enough.  Note that you'd have to make sure mod_pagespeed
was able to initiate fetches to the origin in that case.

What's the use-case for your forward proxy, out of curiosity?

**Comment by mjkeating** [comment]
The use case is simply to allow Pagespeed optimization for our mobile users over the internet - as a DCP.  Right now we are going for maximum compression.

We were using CoreFitlers (urls were definately being modified by pagespeed) and got the same behavior with images - three passes no matter what we set for RewriteDeadlinePerFlushMs. We found that OptimizeForBandwidth improved both compression and page breakage, so we are now using that.

But if there is a particular configuration we could try to get that 'first pass' compression to work, we would certainly like to try it out.

Many thanks, again, for your reply.

**Comment by jmarantz** [comment]
I thought about this some more.  I think mod_pagespeed can be changed to
support single-pass optimization, but there's a price, especially in a
forward proxy.

Let's say you browse to a site with 50 large images, and each one of them
takes 200ms of CPU time to optimize.  A single CPU will be busy for 10
seconds optimizing those images, blocking the full page until its done, but
saving bandwidth.  If your server is acting as a forward proxy for multiple
concurrent active users, I think there would be situations where the
delivery of images would be blocked for tens of seconds.

Is that acceptable for your use-case?  Do you have enough CPUs to share the
load and keep up with it?  If this is just a private VM you are using to
accelerate your own web experience that should be OK.  If it's shared then
you need to scale up the CPUs.  I'd be interested in talking more about
what your environment is.

Note that the same image-optimization delays occur in mod_pagespeed today:
the module's rewrite queue can back up and take a long time to drain. But
in the meantime, unoptimized resources will be served to clients so they
don't suffer large latencies during background optimization.  This makes
sense in an origin or reverse-proxy where a single site is being optimized
and will eventually be completed and saved to cache.  But in a
forward-proxy that kind of cache-hit rate is unlikely, and there could be
ever-increasing delays if we block serving resources until after they are
optimized.

Having said that, as a user of a mobile phone running Chrome, you can get
exactly this behavior for http sites by enabling the Chrome Data Reduction
Proxy https://developer.chrome.com/multidevice/data-compression.  In fact
the image optimizations you get with this feature use the same code as
PageSpeed.   This is possible because Google can throw a lot of servers at
the problem.

**Comment by hillsp** [comment]
We did have a question about the importance of having the very first request return a compressed object. In a real-world scenario (especially once jmarantz' fix lands) no more than a handful of requests would get unoptimised content. Is there a reason you are particularly sensitive to that, or are you just trying to do right by your users?

**Comment by mjkeating** [comment]
We are just trying to give our users as much optimized (compressed) content as we can. Since this is a DCP and not a local web site, many users will run into pages that haven't been optimized and cached - either because it's a new page for the DCP or because the file/image has a short cache lifetime.  

It would be nice to have the config option of increasing the chance for first-pass compression (perhaps an adjustable setting), and then it would be on us to provide the CPU horsepower necessary for an acceptable user experience.  In other words, we may not guarantee first pass compression, but it could happen - and happen more frequently if configured to do so.

Again, many thanks for your attention :)

**Comment by liweing** [comment]
is there any way to set a different & shorter expiration header for the unoptimized resource which is served on the first requesting ?
If yes, the CND or user side will cache the unoptimized resource for a short time, then we can work around the issue.

**Comment by ianneub** [comment]
We would greatly appreciate the addition of @liweing 's suggestion. We have a CDN sitting in front of PageSpeed. It would make a lot of sense for PageSpeed to be able to control the max-age of URLs that are not 100% optimized yet.

EDIT: This would be great also for the HTML pages that are being modified by PageSpeed. Making it easier to use a CDN for whole site caching.

**Comment by lbc2800** [comment]
Is it the shorter expiration workaround that is being added to the 2.0 milestone, or a fix for the originally reported problem (allow image compression on first pass)? 

I'd like to express support for enabling image compression to be configurably performed on the first pass (up to the configurable rewrite deadline, of course). In markets where bandwidth is very expensive - such as markets with lower income, predominantly still on 2G/3G and/or have ultra high roaming costs - end users would be willing to wait a few more seconds. This tradeoff is worthwhile since it means the ability to load more web pages (sometimes by up to 2X, depending on content). These are also the markets where most (>85%) of the users are on mobile and it makes sense to tune pagespeed for the least amount of bandwidth. 

While Chrome has its own workaround for this issue, many of our users are accessing content through embedded webviews and also often using other browsers.
